Fixes

ooreilly opened the pull request:
This PR introduces the following fixes:
* Resolve #12 
* Prevent DM sync issues by changing `ngsl` from `8` to `4`
* Fix parallel inconsistency when force is applied at the interface of two mpi partitions

Recommendations:
Please check that source propagates correctly across the DM overlap zone. Run two test simulations:
1. Place one source in the top block near the overlap zone. 
2. Place one source in the bottom block near the overlap zone.
Check that it in each case that waves can propagate across the overlap zone with causing numerical artifacts.

ooreilly committed:
Add GDB.

ooreilly committed:
Set CUDA compile flags.

ooreilly committed:
Fix segmentation fault when using point forces on multiple GPUs.

ooreilly committed:
Fix force parallel bug.

ooreilly committed:
Fix ngsl.

ooreilly committed:
Fix repeating source.

ooreilly committed:
Add old CUDA print functions.

ooreilly committed:
Update kernel.cu to fix DM sync issues.

ooreilly committed:
Fix three types of parallel bugs in the source and receiver implementation.
1. Synchronization issues due to overlapping sources (atomics enabled by default)
2. Inconsistencies caused by changing ngsl in previous patch
3. Inconsistencies when using the DM.

commit 30a250af75a1802cf59b31d238e4a51874b9a3de
Author: Ossian O'Reilly <ossian.oreilly@gmail.com>
Date:   Thu Oct 1 10:44:24 2020 -0700

    fix test.

commit 609795374d4f755e9b6f6dfe8d9c822fa3085780
Author: Ossian O'Reilly <ossian.oreilly@gmail.com>
Date:   Thu Oct 1 10:39:25 2020 -0700

    add option to toggle partitioning mechanism depending on if it is a source or receiver.

commit a28d36ecb6d9889fff87f6cb5079acc187244c97
Author: Ossian O'Reilly <ossian.oreilly@gmail.com>
Date:   Wed Sep 30 23:37:19 2020 -0700

    Fix DM source receiver parallel bug. Now, mpi-distribute test is broken.

commit 2ba3fca0f26dd2d21ab7760d3629b5de7475f44c
Author: Ossian O'Reilly <ossian.oreilly@gmail.com>
Date:   Wed Sep 30 22:33:20 2020 -0700

    Modify grid bounds check used by source distribution (increase width by 2 points in each direction). 100 x 100 sources work to rounding error. Exact results if interpolation is disabled because then all the overlapping sources add the same value and then the order doesn't matter.

commit 9fdbad904c5891658b7bda1ca21a753a31eb534f
Author: Ossian O'Reilly <ossian.oreilly@gmail.com>
Date:   Wed Sep 30 20:39:14 2020 -0700

    fix parallel bugs that breaks the source implementation on a single grid and even when interpolation is disabled. This point may have introduced parallel bugs in the receivers. Interpolation not yet working.

commit 7cf4995b41d431ccaa2f4aca07f49fd52b39d1e8
Author: Ossian O'Reilly <ossian.oreilly@gmail.com>
Date:   Wed Sep 30 17:13:32 2020 -0700

    add atomic operations to sources to prevent parallel sync issues when the sources are overlapping.

ooreilly committed:
add print statements for outputting grid numbers.

ooreilly committed:
rewrite distribute function so that it is compatible with multiple blocks.

ooreilly committed:
remove print statements and add some comments to explain certain steps in the source partitioning function.

ooreilly committed:
fix broken test (update client call).

ooreilly committed:
fix source interpolation in the curvilinear kernel.

ooreilly committed:
fix block size for original AWP.

hzfmer reviewed (changes requested):
I heard duplicate or overlapping subfaults issue is solved in this source read approach. I am not sure how/in which commit it was done. Hopefully the change I pointed out won't affect that (I did some preliminary tests, which seems promising) .

There is, however, another bug, that if duplicate receivers are provided, the code freezes at writing the first receiver output  file (it doesn't finish the first "x" output file, not even create "y" and "z" files).

hzfmer committed:
fix a bug in PR-13 to assign receiver/source indices in each partition

ooreilly committed:
add zhat arg to source functions.

ooreilly committed:
fix source amplitude errors for sources placed near the boundary at the cell-centered grid points.

ooreilly committed:
Fix wrong quadrature weights.

ooreilly committed:
change step to size_t in buffer.

ooreilly committed:
fix integer overflow in mpi_io_indexed.

ooreilly committed:
enable debugging statements and fix force boundary issue for a single point with no interpolation.

ooreilly committed:
fix source in the x-direction.

ooreilly committed:
fix force in the y direction.

ooreilly committed:
add partitioning interface, but this change breaks the gaussian hill test.

ooreilly committed:
remove unused macro.

ooreilly committed:
fix output of density.

ooreilly committed:
remove unused grids.

ooreilly committed:
rename partitioning functions and update partitioning bounds to minimize the amount of overlap.

ooreilly committed:
remove debug statements.

ooreilly committed:
Fix SGT output.

ooreilly committed:
add optimized stress kernel to interior region.

ooreilly committed:
add optimized kernel to left boundary.

ooreilly committed:
add optimized stress kernel to right boundary.

ooreilly committed:
forgot to add kernel :)

ooreilly committed:
add missing kernel :)

ooreilly committed:
Fix source placement (#16)

DM compatible source and receiver placement and test. Convergence test and fixes to some broken tests. Point force implementation for the original AWP boundary condition. 

* fix incorrect placement of density and incorrect interpolation of density in force computation.

* begin mms module.

* add mms files.

* fix density interpolation.

* comment out work in progress code.

* override material properties.

* initialize mms solution.

* add forcing functions.

* Use plane wave solutions.

* remove unused libraries, duplicate error header file, and receiver input
file to mms module.

* delete unused kernels.

* remove block size from definitions.h and remove this dependency from some of the files.

* update mms so that uses both P and S plane waves.

* Fix kernel generator so that it runs with updated sympy version, and fix off by one errors in density interpolation.

* Fix off by 0.5h error for source/receiver placement of components that
are located at the cell-centered grid points in the z-direction. This
fix only applies to the DM.

* start adding point force suppport for the traditional bc.

* fix placement of receivers on the free surface for regular AWP.

* add surface force to original AWP. doesn't pass the reciprocity test.

* Implement point forces in original AWP using stress perturbations. Passes the reciprocity test.

* apply symmetry-based boundary conditions, and add body force.

* start reintroducing convergence and energy tests.

* interior convergence for the velocity field is working.

* Cartesian velocity convergence tests shows second order convergence. Added option to enable/disable application of bc.

* get interior stress convergence working.

* convergence test might be working for a cartesian geometry. It is hard to tell in single precision.

* add option to have double-precision topography files.

* remove double-precision support because it doesn't work.

* put the truncation error test as part of the unit test. Document the test.

* add force extrapolation.

* update grid printing.

* remove broken test.

* restore original AWP FS implementation.

* add accuracy test data.

* set c++ standard.

* apply most recent mms changes.

* add new grid implementation that maps the global z coordinate to the local block coordinate.

* add arg to grid_fill1 that corrects for AWP's (-,+,+) choice. Currently,
the correction is not applied.

* remove shift in x-direction. Causes several tests to fail.

* fix convergence test.

* fix broken interpolation test.

* fix broken grid_3d test.

* fix broken metrics tests.

* update dm source/receiver test so that it checks all components.

* remove module grid_new.

* change grid number identification process.

* change to default AWP force implementation.

* restore find_grid_number

* remove old arch flags.

Co-authored-by: Ossian O'Reilly <ooreilly@usc.edu>

ooreilly committed:
Fix mapping (#17)

Modifies the mapping in the overlap zone such that it is compatible with the DM.

* fix incorrect placement of density and incorrect interpolation of density in force computation.

* begin mms module.

* add mms files.

* fix density interpolation.

* comment out work in progress code.

* override material properties.

* initialize mms solution.

* add forcing functions.

* Use plane wave solutions.

* remove unused libraries, duplicate error header file, and receiver input
file to mms module.

* delete unused kernels.

* remove block size from definitions.h and remove this dependency from some of the files.

* update mms so that uses both P and S plane waves.

* Fix kernel generator so that it runs with updated sympy version, and fix off by one errors in density interpolation.

* Fix off by 0.5h error for source/receiver placement of components that
are located at the cell-centered grid points in the z-direction. This
fix only applies to the DM.

* start adding point force suppport for the traditional bc.

* fix placement of receivers on the free surface for regular AWP.

* add surface force to original AWP. doesn't pass the reciprocity test.

* Implement point forces in original AWP using stress perturbations. Passes the reciprocity test.

* apply symmetry-based boundary conditions, and add body force.

* start reintroducing convergence and energy tests.

* interior convergence for the velocity field is working.

* Cartesian velocity convergence tests shows second order convergence. Added option to enable/disable application of bc.

* get interior stress convergence working.

* convergence test might be working for a cartesian geometry. It is hard to tell in single precision.

* add option to have double-precision topography files.

* remove double-precision support because it doesn't work.

* put the truncation error test as part of the unit test. Document the test.

* add force extrapolation.

* update grid printing.

* remove broken test.

* restore original AWP FS implementation.

* add accuracy test data.

* set c++ standard.

* apply most recent mms changes.

* add new grid implementation that maps the global z coordinate to the local block coordinate.

* add arg to grid_fill1 that corrects for AWP's (-,+,+) choice. Currently,
the correction is not applied.

* remove shift in x-direction. Causes several tests to fail.

* fix convergence test.

* fix broken interpolation test.

* fix broken grid_3d test.

* fix broken metrics tests.

* update dm source/receiver test so that it checks all components.

* remove module grid_new.

* change grid number identification process.

* change to default AWP force implementation.

* restore find_grid_number

* remove old arch flags.

* fix mms so that it works with grid stretching.

* reintroduce mapping so that grid stretching function achieves zero at the overlap zone.

* add mapping to vx.

* add mapping to vy.

* add mapping to xz, yz.

* change overlap zone point.

* add mapping to velocity buffers.

Co-authored-by: Ossian O'Reilly <ooreilly@usc.edu>

ooreilly committed:
Grid stretch (#18)

* add new command line options.

* add mapping module.

* add mapping object to topography.

* apply mapping in topography module.

* rename mapping functions.

* add mapping inversion that determines the parameter space coordinate given the physical coordinate.

* fix DM grid stretching problem.

ooreilly committed:
Update grid writing tool (#19)

* fix serial reader test.

* remove print statements.

* fix write grid tool.

* add DM and nonuniform grid stretching compatibility to the curvilinear grid writing tool.

* set block size to its correct size.

* add top and bottom grid spacing parameters.

ooreilly committed:
Fix write material property info (#20)

ooreilly committed:
Fix unit tests (#21)

* start fixing convergence test.

* update convergence test to account for modified mapping.

* fix mapping test.

* remove print statement.

* remove broken test.

ooreilly committed:
rename project. (#22)

ooreilly committed:
Update write_grid docs. (#23)

ooreilly committed:
Energy rate (#24)

Add energy rate output.

* add energy command line args and empty module.

* initialize energy and write dummy output.

* add reduction kernel.

* add all fields.

* energy rate without boundary treatment.

* energy conservation works for homogeneous material properties, no topography, no exterior boundaries.

* add jacobians.

* add random density.

* add shear modulus.

* add lambda parameter.

* add mpi reduce.

* fix off by one error in density interpolation in the x-direction.

* add time vector.

ooreilly committed:
disable energy module when not in use. (#25)

ooreilly committed:
fix incorrect placement of source/receivers in the y-direction and point force parallel consistency issue (#26)

* fix incorrect placement of source/receivers in the y-direction when using multiple grid blocks.

This fix Shifts sources/receivers in the y-direction by one grid spacing of the previous grid: y = y_prev - h_prev.

* remove inactive code in source initialization.

* Refactor source distribution (#27)

* change y-axis for DM blocks instead of moving source/recv.

* init source distribution.

* modify partitioning of receivers so that it is easier to understand.

* modify partitioning of sgts.

* replace old partitioning kernels with new ones that uses grid indices instead of coordinates.

* refactor distribute.

* add grid block number to data init function so that it correctly generates y-values for DM blocks.

* delete old partition functions.

* cleanup partitioning functions and add guard statements to source
kernels.

* remove guard statements.

* add default case.

* update comment.

* Fix parallel bug in point force insertion (#28)

* add error code handling for mapping module

* Fix parallel bug in point force insertion.

* clean-up error handling in mapping.

* Fix point force guards for Cartesian kernel and disable debugging output.

* Refactor source distribution (#29)

* change y-axis for DM blocks instead of moving source/recv.

* init source distribution.

* modify partitioning of receivers so that it is easier to understand.

* modify partitioning of sgts.

* replace old partitioning kernels with new ones that uses grid indices instead of coordinates.

* refactor distribute.

* add grid block number to data init function so that it correctly generates y-values for DM blocks.

* delete old partition functions.

* cleanup partitioning functions and add guard statements to source
kernels.

* remove guard statements.

* add default case.

* update comment.

* add error code handling.

* debug partitioning.

* Fix parallel bug in point force insertion.

The point force must be applied before the front/back send/recv kernels
execute.

* clean-up error handling in mapping.

* Fix point force guards for Cartesian kernel and disable debugging output.

* fix bounds guards for point force kernels.

ooreilly committed:
Fix dm overlap placement (#30)

* force source/recvs in the overlap zone to be placed on the coarser grid.

* cleanup.

* fix find_grid_number so that it maps surface coordinates to the top block.

deanrockit committed:
Redesign cerjan function so that ND is no longer bounded by min(nxt,nyt)

deanrockit committed:
Fix write_grid tool so that it works with multiple processors. Fix an issue in write_grid tool with allocation of memory.

deanrockit committed:
* Adding new user input flags specifying Q model parameters and caps on velocity (QSI, QPQSR, MAXVPVSR, VMIN, VMAX, and DMIN).

deanrockit committed:
Delete release folder

deanrockit committed:
Modify velbuffer function to avoid zero values in Vz at the surface when topography is turned on.

deanrockit committed:
Merge pull request #31 from SCECcode/new_cerjan

Redesign cerjan function so that ND is no longer bounded by min(nxt,nyt)

Commit:
Remove textures, add compile scripts for summit,ls6,vista

akashpalla committed:
Change param type of Alloc1D from int to long

akashpalla committed:
Remove mesh_clean.c from compile scripts

deanrockit committed:
Merge pull request #34 from akashpalla/cuda12+

Remove textures for CUDA 12+, Add compile scripts
